### DISCUSSION TITLE
Overrides on different sytems?

### DISCUSSION THREAD

**Discussion by TRPB** [nix-community/nixvim#2372] [General]
I'd like to set a few different global variables on different systems (scaling, some paths configured in vim.g). This might be a documentation issue but I couldn't work out how to do it.

When loading the flake, it would be nice if I could override `vim.g.foo` variables. 

I think we'd need to do something like

```
{

  inputs = {

    nixpkgs.url = "github:nixos/nixpkgs/release-24.05";
    nixvim.url = "github:account/repo";
  };

  outputs =
    {
      self,
      nixpkgs,
      nixvim,
      ...
    }@inputs:
    let
      pkgs = nixpkgs.legacyPackages.x86_64-linux;
    in
    {
      nixosConfigurations.systemname = nixpkgs.lib.nixosSystem {
        system = "x86_64-linux";
        modules = [
          nixvim.modules.nixvimConfig
          ./systemconfig.nix
        ];
      };
}
```

Then in a module call `nixvim.settings.g.variablename = 'x';` 

I had a look through the code, it does export some modules but I don't think I can easily override settings. 

I'd also like to do this on a machine with a different using `nix run` but I assume that's a different challenge  (or home-manager but I don't know how to import flakes into the home manager config)

**Comment by traxys** [comment]
You could approach that by multiple means:
- If you use the standalone nixvim module the resulting package exposes a `.extend { .... }` function that allows to apply additional settings, so you could condition to the system
- You can use multiple nix files, one for your "common" code, and then different ones for each system. You can the include them in the `modules`
- You can use `some.nixvim.option = mkIf (pkgs.stdenv.isAaarch64) { ..... }` to enable something only for aarch64 systems

**Comment by MattSturgeon** [comment]
IMO it's best to keep this kinda logic internal to modual eval whenever possible, so I would recomend using `pkgs.stdenv` for this.

To expand on https://github.com/nix-community/nixvim/discussions/2372#discussioncomment-10857905, you can leverage several attrs:

```
pkgs.stdenv.__bootPackages            pkgs.stdenv.extraBuildInputs          pkgs.stdenv.isFreeBSD                 pkgs.stdenv.overrideDerivation
pkgs.stdenv.__extraImpureHostDeps     pkgs.stdenv.extraNativeBuildInputs    pkgs.stdenv.isLinux                   pkgs.stdenv.overrides
pkgs.stdenv.__hatPackages             pkgs.stdenv.extraSandboxProfile       pkgs.stdenv.isMips                    pkgs.stdenv.passthru
pkgs.stdenv.all                       pkgs.stdenv.fetchurlBoot              pkgs.stdenv.isOpenBSD                 pkgs.stdenv.preHook
pkgs.stdenv.allowedRequisites         pkgs.stdenv.hasCC                     pkgs.stdenv.isSunOS                   pkgs.stdenv.setup
pkgs.stdenv.args                      pkgs.stdenv.hostPlatform              pkgs.stdenv.isi686                    pkgs.stdenv.shell
pkgs.stdenv.bootstrapTools            pkgs.stdenv.initialPath               pkgs.stdenv.isx86_32                  pkgs.stdenv.shellDryRun
pkgs.stdenv.buildPlatform             pkgs.stdenv.is32bit                   pkgs.stdenv.isx86_64                  pkgs.stdenv.shellPackage
pkgs.stdenv.builder                   pkgs.stdenv.is64bit                   pkgs.stdenv.meta                      pkgs.stdenv.system
pkgs.stdenv.cc                        pkgs.stdenv.isAarch32                 pkgs.stdenv.mkDerivation              pkgs.stdenv.targetPlatform
pkgs.stdenv.defaultBuildInputs        pkgs.stdenv.isAarch64                 pkgs.stdenv.name                      pkgs.stdenv.tests
pkgs.stdenv.defaultNativeBuildInputs  pkgs.stdenv.isBSD                     pkgs.stdenv.out                       pkgs.stdenv.type
pkgs.stdenv.disallowedRequisites      pkgs.stdenv.isBigEndian               pkgs.stdenv.outPath
pkgs.stdenv.drvAttrs                  pkgs.stdenv.isCygwin                  pkgs.stdenv.outputName
pkgs.stdenv.drvPath                   pkgs.stdenv.isDarwin                  pkgs.stdenv.override
```

In particular, `hostPlatform.system`[^1] will tell you the system string for the targeted platform. In addition there's several `is*` helpers, such as `isDarwin`, `isLinux`, `isx86_64`, `is64bit`, etc.

[^1]: using `pkgs.stdenv.hostPlatform.system` is preferred over directly using `pkgs.stdenv.system` or `pkgs.system` for several reasons, in particular it allows for cross-compilation.
